cidc_api/models/templates/sync_schemas.py

```diff
@@ -0,0 +1,226 @@
+        errors.extend(new_errors)
+        if new_trial is None:
+            continue
+
+        errors.extend(_sync_collection_events(trial, new_trial, session))
+        errors.extend(_sync_cohorts(trial, new_trial, session))
+        errors.extend(_sync_shipments(trial, new_trial, session))
+
+        # Since we'll be referencing these for Participant and Sample
+        session.flush()
+
+        errors.extend(_sync_participants_and_samples(trial, new_trial, session))
+
+    # final flush before commit / rollback
+    session.flush()
+
+    if dry_run or len(errors):
+        session.rollback()
+    else:
+        session.commit()
+    session.close()
+
+    return errors
+
+
+def _generate_new_trial(
+    trial: TrialMetadata, session: Session
+) -> Tuple[List[Exception], Optional[ClinicalTrial]]:
+    """
+    Given a TrialMetadata, set up a new ClinicalTrial.
+    Returns one of:
+        [Exception()], None
+        [], ClinicalTrial()
+    """
+    try:
+        new_trial = session.merge(
+            ClinicalTrial(
+                protocol_identifier=trial.trial_id,
+                **{
+                    c.name: trial.metadata_json[c.name]
+                    for c in ClinicalTrial.__table__.columns
+                    if c.name != "protocol_identifier" and c.name in trial.metadata_json
+                },
+            )
+        )
+        session.flush()
+        return [], new_trial
+
+    except Exception as e:
+        return [e], None
+
+
+def _sync_collection_events(
+    trial: TrialMetadata, new_trial: ClinicalTrial, session: Session
+) -> List[Exception]:
+    """
+    Given corresponding TrialMetadata and ClinicalTrial instances, copy over collection events.
+    In JSON, stored as list[string] at /allowed_collection_event_names.
+    """
+    errors = []
+    for event_name in trial.metadata_json["allowed_collection_event_names"]:
+        try:
+            session.merge(
+                CollectionEvent(
+                    event_name=event_name, trial_id=new_trial.protocol_identifier
+                )
+            )
+        except Exception as e:
+            errors.append(e)
+
+    return errors
+
+
+def _sync_cohorts(
+    trial: TrialMetadata, new_trial: ClinicalTrial, session: Session
+) -> List[Exception]:
+    """
+    Given corresponding TrialMetadata and ClinicalTrial instances, copy over cohorts.
+    In JSON, stored as list[string] at /allowed_cohort_names.
+    """
+    errors = []
+    for cohort_name in trial.metadata_json["allowed_cohort_names"]:
+        try:
+            session.merge(
+                Cohort(cohort_name=cohort_name, trial_id=new_trial.protocol_identifier)
+            )
+        except Exception as e:
+            errors.append(e)
+
+    return errors
+
+
+def _sync_shipments(
+    trial: TrialMetadata, new_trial: ClinicalTrial, session: Session
+) -> List[Exception]:
+    """
+    Given corresponding TrialMetadata and ClinicalTrial instances, copy over shipments.
```

Comment:
Similar to a previous comment: How do you deal with the trials in-which there are multiple shipping manifest uploads (even if if they are the same). I feel like there probably needs to be some code in this function to address that, or we'll have to have cleaned the database.

Have you tested this sync in somewhere with a full copy of production? @scvannost

Reply:
[`session.merge`](https://docs.sqlalchemy.org/en/14/orm/session_api.html#sqlalchemy.orm.Session.merge) here will try to handle combining the entries using the primary keys, which are [trial_id and manifest_id](https://github.com/CIMAC-CIDC/cidc-api-gae/pull/549/files#diff-b1fbf246c0e7e54177b44319807161904f88c2aa8d0f7d7b5dacab17946138a7R201-R204)